src/libraries/Microsoft.PowerFx.Core/Utils/TextFormatUtils.cs

```diff
@@ -85,21 +97,154 @@ public static bool IsValidFormatArg(string formatString, CultureInfo formatCultu
+
+            //Block "gen"/"general"
+            if (formatStr.IndexOf("gen", StringComparison.OrdinalIgnoreCase) >= 0)
             {
-                // Check if the date time format contains '0's after the seconds specifier, which
-                // is used for fractional seconds - in which case it is valid
-                var formatWithoutZeroSubseconds = _formatWithoutZeroSubsecondsRegex.Replace(textFormatArgs.FormatArg, m => m.Groups[1].Success ? string.Empty : m.Groups[1].Value);
-                textFormatArgs.HasNumericFmt = formatWithoutZeroSubseconds.IndexOfAny(new char[] { '0', '#' }) >= 0;
+                return false;
             }
 
-            if (textFormatArgs.HasDateTimeFmt && textFormatArgs.HasNumericFmt)
+            for (int i = 0; i < formatStr.Length; i++)
             {
-                return false;
+                // Do not allow format string start with '/'
+                if (i == 0 && formatStr[i] == '/')
+                {
+                    return false;
+                }
+
+                if (_numericCharacters.Contains(formatStr[i]))
+                {
+                    // ':' is not allowed between # or 0 (numeric)
+                    if (!textFormatArgs.HasDateTimeFmt && hasColonWithNum)
+                    {
+                        return false;
+                    }
+
+                    textFormatArgs.HasNumericFmt = true;
+
+                    // Use hasNumericCharacters to check if format string has numeric character before group separator or after decimal point.
+                    hasNumericCharacters = true;
+                }
+                else if (_dateTimeCharacters.Contains(formatStr[i]))
+                {
+                    textFormatArgs.HasDateTimeFmt = true;
+                }
+                else if (!textFormatArgs.HasDateTimeFmt && formatStr[i] == ',' && !hasNumericCharacters)
+                {
+                    // If there is no numeric format character before group separator character, then treat it as an escaping character.
+                    formatStr = formatStr.Insert(i, "\\");
+                    i++;
+                }
+                else if (!textFormatArgs.HasDateTimeFmt && formatStr[i] == '.')
+                {
+                    // Reset hasNumericCharacters to false to later check if any numeric character after decimal point.
+                    decimalPointIndex = i;
+                    hasNumericCharacters = false;
+                }
+                else if (_unsupportedCharacters.Contains(formatStr[i]))
+                {
+                    // Block all unsupported characters
+                    return false;
+                }
+                else if (formatStr[i] == ':')
+                {
+                    hasColonWithNum = true;
+                }
+                else if (formatStr[i] == ';')
+                {
+                    // Does not allow number of section are more than 2.
```

Comment:
>// Does not allow number of section are more than 2. [](http://example.com/codeflow?start=20&length=52)

Do we support more than 1 section? Would the results of `Text(12345.6789,"0.00000;#.######")` and `Text(12345.6789,"#.######;0.00000")` be different?

Reply:
Yes, we support upto 3 sections. If only 2 section, the first one will be used for zero (same with excel) as:
> Text(12345.6789,"0.00000;#.######")
"12345.67890"

> Text(12345.6789,"#.######;0.00000")
"12345.6789"